lib/trino-parquet/src/main/java/io/trino/parquet/metadata/ParquetMetadata.java

```diff
@@ -13,32 +13,268 @@
+        checkArgument(offset.isEmpty() && length.isEmpty() || (offset.isPresent() && length.isPresent()),
+                "Both offset and length must be present or absent");
+        this.parquetMetadata = new FileMetadata(readMessageType(), keyValueMetaData(fileMetaData), fileMetaData.getCreated_by());
     }
 
-    public List<BlockMetadata> getBlocks()
+    private static Map<String, String> keyValueMetaData(FileMetaData fileMetaData)
     {
-        return blocks;
+        if (fileMetaData.getKey_value_metadata() == null) {
+            return ImmutableMap.of();
+        }
+        return fileMetaData.getKey_value_metadata().stream().collect(toMap(KeyValue::getKey, KeyValue::getValue));
+    }
+
+    private MessageType readMessageType()
+            throws ParquetCorruptionException
+    {
+        List<SchemaElement> schema = fileMetaData.getSchema();
+        validateParquet(!schema.isEmpty(), dataSourceId, "Schema is empty");
+
+        Iterator<SchemaElement> schemaIterator = schema.iterator();
+        SchemaElement rootSchema = schemaIterator.next();
+        Types.MessageTypeBuilder builder = Types.buildMessage();
+        readTypeSchema(builder, schemaIterator, rootSchema.getNum_children());
+        return builder.named(rootSchema.name);
     }
 
     public FileMetadata getFileMetaData()
     {
-        return fileMetaData;
+        return parquetMetadata;
+    }
+
+    public List<BlockMetadata> getBlocks(Collection<ColumnDescriptor> columnDescriptors)
+    {
+        Set<ColumnPath> paths = columnDescriptors.stream()
+                .map(ColumnDescriptor::getPath)
+                .map(ColumnPath::get)
+                .collect(toImmutableSet());
+
+        try {
+            return buildBlocks(paths);
+        }
+        catch (ParquetCorruptionException e) {
+            throw new TrinoException(CORRUPT_PAGE, e);
+        }
+    }
+
+    public List<BlockMetadata> getBlocks()
+    {
+        return getBlocks(ImmutableSet.of());
+    }
+
+    private List<BlockMetadata> buildBlocks(Set<ColumnPath> paths)
+            throws ParquetCorruptionException
+    {
+        List<SchemaElement> schema = fileMetaData.getSchema();
+        validateParquet(!schema.isEmpty(), dataSourceId, "Schema is empty");
+
+        MessageType messageType = readParquetSchema(schema);
+        ImmutableList.Builder<BlockMetadata> builder = ImmutableList.builder();
+        List<RowGroup> rowGroups = fileMetaData.getRow_groups();
+        if (rowGroups != null) {
+            for (int i = 0; i < rowGroups.size(); i++) {
+                RowGroup rowGroup = rowGroups.get(i);
+                List<ColumnChunk> columns = rowGroup.getColumns();
+                validateParquet(!columns.isEmpty(), dataSourceId, "No columns in row group: %s", rowGroup);
+                String filePath = columns.getFirst().getFile_path();
+
+                if (offset.isPresent() && length.isPresent() && rowGroup.isSetFile_offset()) {
```

Comment:
I don't get why we needed to refactor the code for introducing this optimization, can you undo the refactoring and make this change in the existing createParquetMetadata method in MetadataReader ?
That will make it easier to see what changed

Reply:
Refactored the ParquetMetadata because there are 3 steps:
1. read Parquet metadata to get schema (in parquet)
2. use schema to find required columns (in catalog like Iceberg)
3. read blocks metadata on required columns (in parquet)

ParquetMetadata class now holds the metadata from step 1 and do lazy calculation on step 3. 

I can change split the commit into one with no-op refactor, and another with optimization, if that can make review easier.

Reply:
Yes, please split this into multiple commits